compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -2138,32 +2147,70 @@ fn prefetch_mir(tcx: TyCtxt<'_>) {
 // will allow us to slice the metadata to the precise length that we just
 // generated regardless of trailing bytes that end up in it.
 
-#[derive(Encodable, Decodable)]
 pub struct EncodedMetadata {
-    raw_data: Vec<u8>,
+    mmap: Option<Mmap>,
+    // We need to carry MaybeTempDir to avoid deleting the temporary
+    // directory while accessing the Mmap.
+    _temp_dir: Option<MaybeTempDir>,
```

Comment:
Do we need a manual `Drop` impl to ensure these are dropped in the right order?  Or is the declaration order sufficient?  This deserves a comment.